fix(common): `Location` does not support base href containing `origin`

constantant opened the pull request:
In case `APP_BASE_HREF` is set including `origin` the further usage of it might cause failure 

e.g.
If an app is placed on `https://example.com` and bundles are on `https://cdn-example.com` you have to set `APP_BASE_HREF` up as `https://example.com/` and build the app with `--base-href` as `https://cdn-example.com/` but it does not work because of the bug

Fixes #48175

## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
`Location` does not support base href containing `origin` (only base hrefs with just the path are supported)
Issue Number: 48175

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

atscott reviewed (changes requested):
nit, but could you update the header of the commit to be the fix rather than a description of the old problem: "fix(common): Update `Location` to support base href containing origin"

Also, could you add a test or a few for this case in `packages/common/test/location/location_spec.ts`?

Lastly, if you're up for it, do you want to open another PR to fix https://github.com/angular/angular/issues/45744?

atscott commented on packages/common/src/location/location.ts:
```diff
@@ -68,7 +68,7 @@ export class Location implements OnDestroy {
   constructor(locationStrategy: LocationStrategy) {
     this._locationStrategy = locationStrategy;
     const browserBaseHref = this._locationStrategy.getBaseHref();
-    this._baseHref = stripTrailingSlash(_stripIndexHtml(browserBaseHref));
+    this._baseHref = _stripOrigin(stripTrailingSlash(_stripIndexHtml(browserBaseHref)));
```
```suggestion
    const baseHref = this._locationStrategy.getBaseHref();
    this._basePath = _stripOrigin(stripTrailingSlash(_stripIndexHtml(baseHref)));
```

Also a nit, but what do you think about updating some of these variable names? `browserBaseHref` is a bit misleading because it makes me think it's the `base href` from the browser, but it could be from APP_BASE_HREF. Also `this._baseHref` is misleading because it's _not_ the same as `baseHref` because we only want to keep the `path` part due to how it's used in this class.

constantant commented:
@atscott could you help to solve the FAIL in the pipeline
```
FAIL: //packages/common:common_api (see /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/execroot/angular/bazel-out/k8-fastbuild/testlogs/packages/common/common_api/test_attempts/attempt_1.log)
[10,978 / 10,979] 365 / 366 tests; Testing //packages/common:common_api; 14s remote
FAIL: //packages/common:common_api (see /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/execroot/angular/bazel-out/k8-fastbuild/testlogs/packages/common/common_api/test.log)

FAILED: //packages/common:common_api (Summary)
      /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/execroot/angular/bazel-out/k8-fastbuild/testlogs/packages/common/common_api/test.log
      /home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/execroot/angular/bazel-out/k8-fastbuild/testlogs/packages/common/common_api/test_attempts/attempt_1.log
INFO: From Testing //packages/common:common_api:
==================== Test output for //packages/common:common_api:
Analysis will use the bundled TypeScript version 4.6.4
Analysis will use the bundled TypeScript version 4.6.4
Analysis will use the bundled TypeScript version 4.6.4
Analysis will use the bundled TypeScript version 4.6.4
Analysis will use the bundled TypeScript version 4.6.4
The following goldens are outdated:
-  testing/index.md
```
https://app.circleci.com/pipelines/github/angular/angular/54009/workflows/2e7862c9-2e32-4826-aa0b-7eb012278fd5/jobs/1265847?invite=true#step-106-61

I am not good at how to enterpretate this kind of messages :)

atscott commented on goldens/public-api/common/testing/index.md:
```diff
@@ -136,7 +136,7 @@ export class SpyLocation implements Location_2 {
     replaceState(path: string, query?: string, state?: any): void;
     // (undocumented)
-    setBaseHref(url: string): void;
+    setBasePath(url: string): void;
```
Sorry if I lead you down this path, but we can't update the name of a public method outside of a major version and without a migration. Let's revert just this name change.

atscott commented on packages/common/test/location/location_spec.ts:
```diff
@@ -215,4 +215,77 @@ describe('Location Class', () => {
+      const baseHref = getBaseHref(origin, basePath);
+
+      TestBed.configureTestingModule({
+        imports: [CommonModule],
```
```suggestion
```
The `CommonModule` import is not needed in any of these tests.

constantant committed:
fix(common): Update `Location` to support base href containing `origin`

In case `APP_BASE_HREF` is set including `origin` the further usage of it might cause failure

e.g.
If an app is placed on `https://example.com` and bundles are on `https://cdn-example.com` you have to set `APP_BASE_HREF` up as `https://example.com/` and build the app with `--base-href` as `https://cdn-example.com/` but it does not work because of the bug

Fixes #48175

atscott commented on packages/common/test/location/location_spec.ts:
```diff
@@ -215,4 +215,77 @@ describe('Location Class', () => {
+
+      TestBed.configureTestingModule({
+        imports: [CommonModule],
+        providers: [
+          {provide: LocationStrategy, useValue: {getBaseHref: () => baseHref, onPopState() {}}},
+          {provide: Location, useClass: Location, deps: [LocationStrategy]},
+        ]
```
```suggestion
        providers: [{provide: APP_BASE_HREF, useValue: baseHref}]
```
Tests should not mock out more things than necessary because it makes the tests more brittle. In this case, no mocks are necessary.

atscott commented on packages/common/test/location/location_spec.ts:
```diff
@@ -215,4 +215,77 @@ describe('Location Class', () => {
+
+      const location = TestBed.inject(Location);
+
+      expect((location as any)._basePath).toBe(basePath);
```
These tests should not make assertions on the internal properties. Instead, ensure the public method `normalize` or `path` gives you the expected result. Before this change, it would not have worked correctly.

atscott reviewed (approved):
👌 LGTM

atscott commented:
caretaker note: Green TGP

AndrewKushnir commented:
This PR was merged into the repository by commit f8ecc194e93bf9f80af0cb0e77032341bf2f9886.
